scanner/src/main/kotlin/experimental/PackageProvenanceResolver.kt

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2021 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.scanner.experimental
+
+import java.io.IOException
+import java.net.HttpURLConnection
+
+import okhttp3.Request
+
+import org.ossreviewtoolkit.downloader.VersionControlSystem
+import org.ossreviewtoolkit.model.ArtifactProvenance
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.RemoteArtifact
+import org.ossreviewtoolkit.model.RepositoryProvenance
+import org.ossreviewtoolkit.model.SourceCodeOrigin
+import org.ossreviewtoolkit.model.UnknownProvenance
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.utils.OkHttpClientHelper
+import org.ossreviewtoolkit.utils.collectMessagesAsString
+import org.ossreviewtoolkit.utils.createOrtTempDir
+import org.ossreviewtoolkit.utils.log
+import org.ossreviewtoolkit.utils.showStackTrace
+
+/**
+ * The [PackageProvenanceResolver] provides a function to resolve the [Provenance] of a [Package].
+ */
+interface PackageProvenanceResolver {
+    /**
+     * Resolve the [Provenance] of [pkg] based on the provided [sourceCodeOriginPriority]. If no provenance can be
+     * resolved [UnknownProvenance] is returned.
+     */
+    fun resolveProvenance(pkg: Package, sourceCodeOriginPriority: List<SourceCodeOrigin>): Provenance
+}
+
+/**
+ * The default implementation of [PackageProvenanceResolver].
+ */
+class DefaultPackageProvenanceResolver : PackageProvenanceResolver {
+    /**
+     * Resolve the [Provenance] of [pkg] based on the provided [sourceCodeOriginPriority]. For source artifacts it is
+     * verified that the [RemoteArtifact] does exist. For a VCS it is verified that the revision exists. If the revision
+     * provided by the [package][pkg] metadata does not exist or is missing, the function tries to guess the tag based
+     * on the name and version of the [package][pkg].
+     */
+    override fun resolveProvenance(pkg: Package, sourceCodeOriginPriority: List<SourceCodeOrigin>): Provenance {
+        sourceCodeOriginPriority.distinct().forEach { sourceCodeOrigin ->
```

Comment:
Should we make `sourceCodeOriginPriority` a `Set` to be able to spare the `distinct()` here?

Reply:
No, because the order defines the priority, and the iteration order of a `Set` depends on the implementation.

Reply:
Then maybe generalize to use a `Collection` so that *also* sets can be passed in (as practically sets in Kotlin are created via `mutableSetOf()`, which creates a set that maintains the iteration order)? And then I'd also omit the `.distinct()` (which internally also just converts to a mutable set and then back to a list) and leave it to the caller to not pass in any duplicates.

Reply:
I don't understand why this `distinct()` call is bothering you so much. I find it important that the function requires the user to provide the sources sorted by priority. With a set or collection it is possible to accidentally pass the wrong order, e.g. by using the wrong set implementation, with a list not. And why leave it to the caller to make the mistake to pass duplicates if it can be prevented with a simple `distinct()` call?

Btw, the config setting which is what will be passed to this function later on is a list already:
https://github.com/oss-review-toolkit/ort/blob/fd32da954a63ae7ceaf0e1b8d81af7e8718682e5/model/src/main/kotlin/config/DownloaderConfiguration.kt#L33-L36

Reply:
> I don't understand why this `distinct()` call is bothering you so much

Most because its implementation is to inelegant, with its crude conversion to a set and back to a list, creating temporary collections.



> And why leave it to the caller to make the mistake to pass duplicates if it can be prevented with a simple `distinct()` call?

Because passing duplicates actually is no mistake, or? What matters here is the relative order of *different* items in the list; specifying the same item multiple times should not change the outcome, but it's just confusing.

Reply:
I have removed the `distinct()` call but kept the signature.